cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java

```diff
@@ -367,4 +375,83 @@ private HttpResponse doRequest(HttpRequest.Builder requestBuilder) throws Datase
     }
     return builder.addHeader(Constants.Security.Headers.USER_ID, userId);
   }
+
+  private static void logThreadDump() {
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Timeout while making request to dataset service. Thread dump follows:");
+      ThreadMXBean bean = ManagementFactory.getThreadMXBean();
+      for (ThreadInfo threadInfo : bean.dumpAllThreads(true, true)) {
+        LOG.trace("{}", toString(threadInfo));
```

Comment:
can we have all of the threadinfos in one log message, instead of split across multiple log messages?
This will make it easier to correlate them.
For instance, there may be two threads that are both calling the `logThreadDump` method, and their log messages get intertwined.

Reply:
good point, done.